lib/agent0/agent0/hyperdrive/policies/lpandarb.py

```diff
@@ -0,0 +1,642 @@
+
+
+def calc_spot_price_local(
+    initial_share_price: FixedPoint,
+    share_reserves: FixedPoint,
+    share_adjustment: FixedPoint,
+    bond_reserves: FixedPoint,
+    time_stretch: FixedPoint,
+) -> FixedPoint:
+    """Calculate spot price.
+
+    Parameters
+    ----------
+    initial_share_price : FixedPoint
+        The initial price of the share.
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    share_adjustment : FixedPoint
+        The amount of share adjustment.
+    bond_reserves : FixedPoint
+        The amount of bond reserves.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The spot price.
+    """
+    effective_share_reserves = share_reserves - share_adjustment
+    return (initial_share_price * effective_share_reserves / bond_reserves) ** time_stretch
+
+
+def calc_apr(
+    share_reserves: FixedPoint,
+    share_adjustment: FixedPoint,
+    bond_reserves: FixedPoint,
+    initial_share_price: FixedPoint,
+    position_duration_seconds: FixedPoint,
+    time_stretch: FixedPoint,
+) -> FixedPoint:
+    """Calculate APR.
+
+    Parameters
+    ----------
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    share_adjustment : FixedPoint
+        The adjustment for share reserves.
+    bond_reserves : FixedPoint
+        The amount of bond reserves.
+    initial_share_price : FixedPoint
+        The initial price of the share.
+    position_duration_seconds : FixedPoint
+        The duration of the position, in seconds.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The APR.
+    """
+    annualized_time = position_duration_seconds / fp_seconds_in_year
+    spot_price = calc_spot_price_local(
+        initial_share_price, share_reserves, share_adjustment, bond_reserves, time_stretch
+    )
+    return (fp1 - spot_price) / (spot_price * annualized_time)
+
+
+def calc_k(
+    share_price: FixedPoint,
+    initial_share_price: FixedPoint,
+    share_reserves: FixedPoint,
+    bond_reserves: FixedPoint,
+    time_stretch: FixedPoint,
+) -> FixedPoint:
+    """Calculate the AMM invariant.
+
+    Uses the following equation:
+        k_t = (c / mu) * (mu * z) ** (1 - t) + y ** (1 - t)
+
+    Parameters
+    ----------
+    share_price : FixedPoint
+        The price of the share.
+    initial_share_price : FixedPoint
+        The initial price of the share.
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    bond_reserves : FixedPoint
+        The amount of bond reserves.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The AMM invariant.
+    """
+    return (share_price / initial_share_price) * (initial_share_price * share_reserves) ** (
+        fp1 - time_stretch
+    ) + bond_reserves ** (fp1 - time_stretch)
+
+
+def get_shares_in_for_bonds_out(
```

Comment:
I know we went back-and-forth on this, sorry if I'm repeating the discussion. But I am confused why this function can't be replaced with the API `get_in_for_out` + `calculate_fees_out_given_shares_in`?

Reply:
Oh, I think I get it. It is because you're adjusting the pool state in a loop later, and you want to modify the params. So cleaning this up is in line with cleaning up the `_local` functions above -- we need to allow the API endpoints to receive custom pool_info and pool_config (really just pool_info?) values. Perhaps this can be summed up as one big TODO + ticket and all of the issues resolved?

Reply:
I thought we couldn't calculate shares in for bonds out.
That's if get_in_for_out returns a result in the same units as you put in.
If your input is bonds, you get out bonds.
If your input is shares, you get out shares.
That's how I understand it.

Reply:
oh yeah, that reason too

Reply:
`get_in_for_out` has two arguments: `amount_out, shares_out`:

if `shares_out`: # amount_out is in shares
  return bonds_in for given shares_out
else: # amount_out is in bonds
  return shares_in for given bonds_out
  
And the converse is true for `get_out_for_in`, with arguments `amount_in, shares_in`:
if `shares_in`: # amount_in is in shares
  return bonds_out for shares_in
else: # amount_in is in bonds
  return shares_out for bonds_in